cleverhans/experimental/certification/utils.py

```diff
@@ -216,8 +218,11 @@ def tf_lanczos_smallest_eigval(vector_prod_fn,
   q_vectors = tf.TensorArray(
       dtype, size=1, dynamic_size=True, element_shape=(matrix_dim, 1))
 
-  # Create random vector with Euclidean norm 1
-  b = tf.random_normal(shape=(matrix_dim, 1), dtype=dtype)
+  # If start vector is all zeros, make it a random normal vector and run for 1000 iter
+  if tf.reduce_sum(b) == 0:
```

Comment:
summing up and comparing to zero could be fragile. First of all it's possible that legitimate non-zero b will sum up to 0. Also generally speaking most of floating point computations introduce errors. I'm not sure whether summing up bunch of zeros could be non zero, but better to avoid it just in case.

I would recommend you to do one of the following:

1. compare `b` to None, if it's None then you create new random vector. If it's non None then you use it directly.
   The disadvantage is that you have to call `tf_lanczos_smallest_eigval` twice, one time with b=None and another with b not None, thus you will have two subgraphs.

2. Provide boolean flag to this function and based on value of this flag re-initialize b:
  
```
def tf_lanczos_smallest_eigval(... , use_warm_start, ...):
  
  if use_warm_start == False:   # may need to try few variations of the condition to make autograph happy
    b = tf.random_normal(...)
```

Keep in mind that such `use_warm_start` flag could be Tensor of type `tf.bool`, i.e. `tf.placeholder(tf.bool, shape=())`

3. If you want to actually compare norm of `b` with something rather than having another boolean argument or None, then you should use `tf.norm(b) < TOLERANCE`.

Reply:
I will do #3 because if the norm is anyways less than a tolerance, then we should not be using that vector